File: generated_codes/chat_agents/chat-agents-java/skiplist/chatgpt_java/output6_skiplist.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {
private static final int MAX_LEVEL = 32;
private final Node head;
private final Node tail;


private static final class Node {
    final int key;
    final int topLevel;
    final AtomicMarkableReference<Node>[] next;

    @SuppressWarnings("unchecked")
    Node(int key, int height) {
        this.key = key;
        this.topLevel = height;
        this.next = (AtomicMarkableReference<Node>[]) new AtomicMarkableReference[height + 1];
        for (int i = 0; i <= height; i++) {
            this.next[i] = new AtomicMarkableReference<>(null, false);
        }
    }
}

public ConcurrentDataStructure() {
    tail = new Node(Integer.MAX_VALUE, MAX_LEVEL);
    head = new Node(Integer.MIN_VALUE, MAX_LEVEL);
    for (int i = 0; i <= MAX_LEVEL; i++) {
        head.next[i].set(tail, false);
    }
}

@Override
public boolean add(int key) {
    int topLevel = randomLevel();
    Node[] preds = new Node[MAX_LEVEL + 1];
    Node[] succs = new Node[MAX_LEVEL + 1];

    while (true) {
        boolean found = find(key, preds, succs);
        if (found) {
            return false;
        }

        Node newNode = new Node(key, topLevel);
        for (int level = 0; level <= topLevel; level++) {
            newNode.next[level].set(succs[level], false);
        }

        Node pred = preds[0];
        Node succ = succs[0];

        if (!pred.next[0].compareAndSet(succ, newNode, false, false)) {
            continue;
        }

        for (int level = 1; level <= topLevel; level++) {
            while (true) {
                pred = preds[level];
                succ = succs[level];

                if (pred.next[level].compareAndSet(succ, newNode, false, false)) {
                    break;
                }

                find(key, preds, succs);
            }
        }

        return true;
    }
}

@Override
public boolean remove(int key) {
    Node victim;
    boolean isMarked;
    boolean[] marked = {false};
    Node[] preds = new Node[MAX_LEVEL + 1];
    Node[] succs = new Node[MAX_LEVEL + 1];

    while (true) {
        boolean found = find(key, preds, succs);
        if (!found) {
            return false;
        }

        victim = succs[0];

        for (int level = victim.topLevel; level >= 1; level--) {
            Node succ = victim.next[level].get(marked);
            while (!marked[0]) {
                victim.next[level].compareAndSet(succ, succ, false, true);
                succ = victim.next[level].get(marked);
            }
        }

        Node succ = victim.next[0].get(marked);
        while (true) {
            isMarked = victim.next[0].compareAndSet(succ, succ, false, true);
            if (isMarked) {
                // Node has been marked
                find(key, preds, succs);
                return true;
            }

            succ = victim.next[0].get(marked);
            if (marked[0]) {
                return false;
            }
        }
    }
}

@Override
public boolean contains(int key) {
    boolean[] marked = {false};
    Node curr = head;

    for (int level = MAX_LEVEL; level >= 0; level--) {
        curr = curr.next[level].getReference();

        while (curr.key < key) {
            curr = curr.next[level].getReference();
        }

        if (curr.key == key) {
            curr.next[0].get(marked);
            return !marked[0];
        }
    }

    return false;
}

private boolean find(int key, Node[] preds, Node[] succs) {
    boolean[] marked = {false};
    boolean snip;
    Node pred;
    Node curr;
    Node succ;

    retry:
    while (true) {
        pred = head;

        for (int level = MAX_LEVEL; level >= 0; level--) {
            curr = pred.next[level].getReference();

            while (true) {
                succ = curr.next[level].get(marked);

                while (marked[0]) {
                    snip = pred.next[level].compareAndSet(curr, succ, false, false);
                    if (!snip) {
                        continue retry;
                    }
                    curr = pred.next[level].getReference();
                    succ = curr.next[level].get(marked);
                }

                if (curr.key < key) {
                    pred = curr;
                    curr = succ;
                } else {
                    break;
                }
            }

            preds[level] = pred;
            succs[level] = curr;
        }

        return curr.key == key;
    }
}

private int randomLevel() {
    int level = 0;
    int rnd = ThreadLocalRandom.current().nextInt();

    while ((rnd & 1) == 1 && level < MAX_LEVEL) {
        level++;
        rnd >>>= 1;
    }

    return level;
}


}

File: generated_codes/chat_agents/chat-agents-cpp/skiplist/claude_sonnet_4.6/output6_skiplist.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <random>

class ConcurrentDataStructure : public SetADT {
    static constexpr int MAX_LEVEL = 16;

    struct Node {
        int val;
        int topLevel;
        std::atomic<Node*> forward[MAX_LEVEL];

        Node(int v, int lvl) : val(v), topLevel(lvl) {
            for (int i = 0; i < MAX_LEVEL; i++)
                forward[i].store(nullptr, std::memory_order_relaxed);
        }
    };

    static bool is_marked_ref(Node* p) { return (reinterpret_cast<uintptr_t>(p) & 1UL) != 0; }
    static Node* get_unmarked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1UL); }
    static Node* get_marked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1UL); }

    Node* head_;
    Node* tail_;
    std::mt19937 rng_;
    std::bernoulli_distribution coin_;

    int randomLevel() {
        int lvl = 1;
        while (lvl < MAX_LEVEL && coin_(rng_)) lvl++;
        return lvl;
    }

    bool find(int key, Node** preds, Node** succs) {
    retry:
        Node* pred = head_;
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            Node* curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
            while (true) {
                Node* succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                while (is_marked_ref(succ_raw)) {
                    Node* snext = get_unmarked_ref(succ_raw);
                    Node* expected = curr;
                    if (!pred->forward[lv].compare_exchange_strong(expected, snext,
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        goto retry;
                    curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
                    succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                }
                if (curr->val < key) { pred = curr; curr = get_unmarked_ref(succ_raw); }
                else break;
            }
            preds[lv] = pred;
            succs[lv] = curr;
        }
        return succs[0]->val == key;
    }

public:
    ConcurrentDataStructure() : rng_(std::random_device{}()), coin_(0.5) {
        head_ = new Node(INT_MIN, MAX_LEVEL);
        tail_ = new Node(INT_MAX, MAX_LEVEL);
        for (int i = 0; i < MAX_LEVEL; i++)
            head_->forward[i].store(tail_, std::memory_order_relaxed);
    }

    ~ConcurrentDataStructure() {
        Node* curr = head_;
        while (curr) {
            Node* next = get_unmarked_ref(curr->forward[0].load(std::memory_order_relaxed));
            delete curr;
            curr = next;
        }
    }

    // Optimistic: no cleanup, pure read path using acquire loads
    bool contains(int key) override {
        Node* curr = get_unmarked_ref(head_->forward[MAX_LEVEL - 1].load(std::memory_order_acquire));
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            while (true) {
                if (is_marked_ref(curr->forward[lv].load(std::memory_order_acquire))) {
                    curr = get_unmarked_ref(curr->forward[lv].load(std::memory_order_acquire));
                    continue;
                }
                if (curr->val < key)
                    curr = get_unmarked_ref(curr->forward[lv].load(std::memory_order_acquire));
                else break;
            }
        }
        return curr->val == key
            && !is_marked_ref(curr->forward[0].load(std::memory_order_acquire));
    }

    bool add(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        int lvl = randomLevel();
        while (true) {
            if (find(key, preds, succs)) return false;
            Node* newNode = new Node(key, lvl);
            for (int i = 0; i < lvl; i++)
                newNode->forward[i].store(succs[i], std::memory_order_relaxed);
            Node* expected = succs[0];
            if (!preds[0]->forward[0].compare_exchange_strong(expected, newNode,
                    std::memory_order_acq_rel, std::memory_order_acquire)) {
                delete newNode; continue;
            }
            for (int i = 1; i < lvl; i++) {
                while (true) {
                    expected = succs[i];
                    if (preds[i]->forward[i].compare_exchange_strong(expected, newNode,
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        break;
                    find(key, preds, succs);
                }
            }
            return true;
        }
    }

    bool remove(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        while (true) {
            if (!find(key, preds, succs)) return false;
            Node* target = succs[0];
            for (int i = target->topLevel - 1; i >= 1; i--) {
                Node* succ = target->forward[i].load(std::memory_order_acquire);
                while (!is_marked_ref(succ))
                    if (target->forward[i].compare_exchange_weak(succ, get_marked_ref(succ),
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        break;
            }
            Node* succ = target->forward[0].load(std::memory_order_acquire);
            while (true) {
                if (is_marked_ref(succ)) return false;
                if (target->forward[0].compare_exchange_strong(succ, get_marked_ref(succ),
                        std::memory_order_acq_rel, std::memory_order_acquire)) {
                    find(key, preds, succs);
                    return true;
                }
            }
        }
    }
};
